compiler/rustc_trait_selection/src/traits/select/mod.rs

```diff
@@ -214,6 +213,122 @@ enum BuiltinImplConditions<'tcx> {
     Ambiguous,
 }
 
+fn tys_may_be_equivalent<'tcx>(oblig_ty: Ty<'tcx>, impl_ty: Ty<'tcx>) -> bool {
+    let k = impl_ty.kind();
+
+    // First, look for obvious equivalences, e.g. the types are equal. The most
+    // common case is to find such an equivalence trivially, e.g. `ty::Bool`
+    // and `ty::Bool`.
+    let mut maybe_equiv = match oblig_ty.kind() {
+        ty::Bool => matches!(k, ty::Bool),
+        ty::Char => matches!(k, ty::Char),
+        ty::Int(ity) => matches!(k, ty::Int(ity2) if ity == ity2),
+        ty::Uint(uty) => matches!(k, ty::Uint(uty2) if uty == uty2),
+        ty::Float(fty) => matches!(k, ty::Float(fty2) if fty == fty2),
+        ty::Adt(def_id, substs1) => match k {
+            ty::Adt(def_id2, substs2) => {
+                def_id == def_id2 && {
+                    // Check for cases like `Foo<3>` vs `Foo<4>`, which are hot
+                    // in crates such as `bitmaps` and `nalgebra`.
+                    let differ_by_single_const_arg = if
+                        substs1.len() == 1
+                        && substs2.len() == 1
+                        && let ty::subst::GenericArgKind::Const(c1) = substs1[0].unpack()
+                        && let ty::subst::GenericArgKind::Const(c2) = substs2[0].unpack()
+                        && let Some(s1) = c1.val().try_to_scalar_int()
+                        && let Some(s2) = c2.val().try_to_scalar_int()
+                        && s1 != s2
+                    {
+                        true
+                    } else {
+                        false
+                    };
+                    !differ_by_single_const_arg
+                }
+            }
+            _ => false,
+        },
+        ty::Str => matches!(k, ty::Str),
+        ty::Array(..) => matches!(k, ty::Array(..)),
+        ty::Slice(..) => matches!(k, ty::Slice(..)),
+        ty::RawPtr(ptr) => matches!(k, ty::RawPtr(ptr2) if ptr.mutbl == ptr2.mutbl),
+        ty::Dynamic(trait_info, ..) => {
```

Comment:
you can also recurse into that one, though iirc that requires you to first sort the existential predicates of the trait object, which might not be worth it